text/0112-server-runtime.md

```diff
@@ -0,0 +1,331 @@
+- `@lwc/engine-dom`: Runtime that can be used to render LWC component trees in a DOM environment. This package is built on top of `@lwc/engine-core`.
+- `@lwc/engine-server`: Runtime that can be used to render LWC component trees as strings. This package is built on top of `@lwc/engine-core`.
+
+The existing `lwc` package stays untouched and will be used to distribute the different versions of the engine. From the developer perspective, the experience should be identical.
+
+**`c/app/app.js`:**
+
+```js
+import { LightningElement } from "lwc";
+
+export default class App extends LightningElement {}
+```
+
+**`client.js`:**
+
+```js
+import { createElement } from "lwc"; // Aliased to @lwc/engine-dom
+import App from "c/app";
+
+const app = createElement("c-app", { is: App });
+document.body.appendChild(app);
+```
+
+**`server.js`:**
+
+```js
+import { createElement, renderToString } from "lwc"; // Aliased to @lwc/engine-server
+import App from "c/app";
+
+const app = createElement("c-app", { is: App });
+const str = renderToString(app);
+
+console.log(str);
+```
+
+### `@lwc/engine-core`
+
+This packages exposes the following platform-agnostic APIs:
+
+- `LightningElement`
+- `api`
+- `track`
+- `readonly`
+- `wire`
+- `setFeatureFlag`
+- `getComponentDef`
+- `isComponentConstructor`
+- `getComponentConstructor`
+- `unwrap`
+- `registerTemplate`
+- `registerComponent`
+- `registerDecorators`
+- `sanitizeAttribute`
+
+The DOM APIs used by the rendered engine are injected by the runtime depending on the environment. A list of all the DOM APIs the engine depends upon can be found in the [DOM APIs usage](#dom-apis-usage) section in the Appendix.
+
+### `@lwc/engine-dom`
+
+This package exposes the following APIs:
+
+- `createElement` + reaction hooks
+- `buildCustomElementConstructor`
+- `isNodeFromTemplate`
+
+This package injects the native DOM APIs into the `@lwc/engine-core` rendering engine.
+
+### `@lwc/engine-server`
+
+This package exposes the following APIs:
+
+- `createElement(name: string, options: { is: typeof LightningElement }): ServerHTMLElement`: This method creates a new LWC component tree. It follows the same signature as the `createElement` API from `@lwc/engine-dom`. Instead of returning a native `HTMLElement`, this method returns a `ServerHTMLElement` with the public properties, aria reflected properties and HTML global attributed.
+- `renderToString(element: ServerHTMLElement): string`: This method synchronously serializes an LWC component tree to HTML. It accepts a single parameter, a `ServerHTMLElement` returned by `createElement` and returns the serialized string.
+
+This package injects mock DOM APIs in the `@lwc/engine-core` rendering engine. Those DOM APIs produce a lightweight DOM structure. This structure can then be serialized into a string containing the HTML serialization of the element's descendants. As described in the Appendix, this package is also in charge of attaching on the global object a mock `CustomEvent`.
```

Comment:
The POC directly serializes the VDOM, without the need of a DOM structure. Why is a lightweight DOM necessary? Isn't that an extra generation step that would impact the performance?

Reply:
Yes, this DOM structure will have a performance impact however there needs to be a DOM structure (even really lightweight) to attach the event listeners and dispatch the DOM events.